### PR TITLE
Implement scoreboard criteria

### PR DESCRIPTION
**Pull request by syldium**
# Description
Scoreboard objectives now use the [new `Criteria` interface](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/d43a1e722f50396694cfaaa96f6059ca533d7b45). This PR also implements entity related methods on scoreboards and objective and fixes a nullability issue with `Objective#displayName()`.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
- Make sure that unit tests are added which test the relevant changes.
- Make sure that the changes follow the existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
- If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/ServerMock.java [hunk 1]
```diff
@@ -1501,8 +1503,7 @@ public void setWarningState(@NotNull WarningState warningState)
 	@Override
 	public @NotNull Criteria getScoreboardCriteria(@NotNull String name)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return this.criteria.computeIfAbsent(name, CriteriaMock::new);
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/ServerMock.java:1507]: Please add a null Check

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.bukkit.scoreboard.RenderType;
+import org.jetbrains.annotations.NotNull;
+
+public class CriteriaMock implements Criteria
+{
+	private final String name;
+	private final boolean readOnly;
+	private final RenderType defaultRenderType;
+
+	public CriteriaMock(String name)
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java]:
```suggestion
	public CriteriaMock(@NotNull String name)
```

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 3]
```diff
@@ -37,31 +42,33 @@ public class ScoreboardMock implements Scoreboard
...
+		return registerNewObjective(name, criteria, displayName, RenderType.INTEGER);
 	}
 
 	@Override
-	public @NotNull Objective registerNewObjective(@NotNull String name, @NotNull Criteria criteria, @Nullable Component displayName, @NotNull RenderType renderType) throws IllegalArgumentException
+	public @NotNull ObjectiveMock registerNewObjective(@NotNull String name, @NotNull Criteria criteria, @Nullable Component displayName, @NotNull RenderType renderType) throws IllegalArgumentException
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		if (this.objectives.containsKey(name))
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java]: Please add Null Checks for the Paramenters

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 4]
```diff
@@ -116,25 +116,26 @@ public ObjectiveMock getObjective(String name) throws IllegalArgumentException
 	@Override
 	public @NotNull Set<Objective> getObjectivesByCriteria(@NotNull Criteria criteria) throws IllegalArgumentException
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		return objectives.values().stream().filter(objective -> objective.getTrackedCriteria().equals(criteria))
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java]: Please add a Null Check

**Review by thelooter — CHANGES_REQUESTED** [review]
Some small Stuff

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package be.seeseemelk.mockbukkit.scoreboard;
+
+import org.bukkit.scoreboard.Criteria;
+import org.bukkit.scoreboard.RenderType;
+import org.jetbrains.annotations.NotNull;
+
+public class CriteriaMock implements Criteria
+{
+	private final String name;
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java]:
```suggestion
{

	private final String name;

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	private final boolean readOnly;
+	private final RenderType defaultRenderType;
+
+	public CriteriaMock(@NotNull String name)
+	{
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/CriteriaMock.java:16]: Missing null check

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 7]
```diff
@@ -13,17 +13,22 @@
...
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import static net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer.legacySection;
+
 public class ScoreboardMock implements Scoreboard
 {
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java:28]:
```suggestion
{

```

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 8]
```diff
@@ -13,17 +13,22 @@
...
+	private static final String OFFLINE_PLAYER_CANNOT_BE_NULL = "OfflinePlayer cannot be null";
+	private static final String ENTITY_CANNOT_BE_NULL = "Entity cannot be null";
+
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java:32]:
```suggestion

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 9]
```diff
@@ -37,31 +42,38 @@ public class ScoreboardMock implements Scoreboard
...
+	{
+		Preconditions.notNull(name, "Objective name cannot be null");
+		Preconditions.notNull(criteria, "Criteria cannot be null");
+		Preconditions.notNull(displayName, "Display name cannot be null");
+		Preconditions.notNull(renderType, "RenderType cannot be null");
+		Preconditions.condition(name.length() <= Short.MAX_VALUE, "The name '" + name + "' is longer than the limit of 32767 characters");
+		if (this.objectives.containsKey(name))
+		{
+			throw new IllegalArgumentException("An objective of name '" + name + "' already exists");
+		}
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java]:
```suggestion
		Preconditions.checkArgument(!this.objectives.containsKey(name), "An objective of name '" + name + "' already exists");

File: src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java [hunk 10]
```diff
@@ -77,64 +89,60 @@ public class ScoreboardMock implements Scoreboard
...
 	}
 
 	@Override
 	public @NotNull Set<Objective> getObjectivesByCriteria(@NotNull Criteria criteria) throws IllegalArgumentException
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.notNull(criteria, "Criteria cannot be null");
+		return objectives.values().stream().filter(objective -> objective.getTrackedCriteria().equals(criteria))
+				.collect(Collectors.toSet());
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/scoreboard/ScoreboardMock.java]:
```suggestion
		return objectives.values().stream()
				.filter(objective -> objective.getTrackedCriteria().equals(criteria))
				.collect(Collectors.toSet());
```

**Review by Insprill — CHANGES_REQUESTED** [review]

**Review by Insprill — APPROVED** [review]
LGTM. Thank you!

**Review by thelooter — APPROVED** [review]

**Pull request merged**